### PR TITLE
Troble

### PR DESCRIPTION
**Pull request by MrTroble**

### CODE REVIEW

File: markdown/02_architecture_constraints.md [hunk 1]
```diff
@@ -1,21 +1,78 @@
...
+| -------------------------- | ----------------------------------------------------------------------------------- | -------------------------------------------------- |
+| Exceptions                 | We don't want our game to crash midway. This adds a lot of unnecessary overhead (5) |                                                    |
+| Cache unfriendly Container | A big risk to miss the L1 cache, which is the fastest cache (3)                     | `std::map, std::list, ...`                         |
+| Per object function        | On a large scale those functions pile up a lot of overhead (5)                      | `texture.create();`                                |
+| Object-Orientation         | The basic nature of object orientation misses the point of data transformation (5)  | `texture.getWidth(); // No optimization guarantee` |
+| DLLs                       | They remove the the optimizers ability to optimize, which is really bad (2)         |                                                    |
+| runtime polymorphism       | This hits cold memory and miss the branch prediction                                | `virtual toString();`                              |
```
Comment (Clazilla) [inline markdown/02_architecture_constraints.md:17]:
```suggestion
| runtime polymorphism       | This hits cold memory and misses the branch prediction                                | `virtual toString();`                              |
```

File: markdown/02_architecture_constraints.md [hunk 2]
```diff
@@ -1,21 +1,78 @@
...
+| Functions in structs or classes | This should only be used by entries on non-critical paths, as it possesses a big risk of potential overhead (4)(5)                                                                  | `struct _Test { const char* getTestName(); };`                |
+| Global variables                | This adds startup overhead, so better not have too much (1)                                                                                                                         | `extern int x  = 0;`                                          |
+| Copy and Swap                   | Try to avoid unnecessary copies and swaps                                                                                                                                           | `std::string name = "Test"; std::string name2 = name; //Copy` |
+| High level abstraction          | Every abstraction comes with a cost, the lower the level the better (4)                                                                                                             | Code generation, Classes, Templates                           |
```
Comment (Clazilla) [inline markdown/02_architecture_constraints.md:29]:
```suggestion
| High level abstraction          | Every abstraction comes with a cost, the lower the level the better (4)                                                                                                             | Code Generation, Classes, Templates                           |
```

File: markdown/02_architecture_constraints.md [hunk 3]
```diff
@@ -1,21 +1,78 @@
...
+| Smart pointer                   | They have hidden costs and threading issues (4)                                                                                                                                     | `std::unique_ptr<int>`                                        |
+
+## Encouraged
+
+In contrast to the lists above, it is strongly recommended to use the following patterns and techniques.
```
Comment (Clazilla) [inline markdown/02_architecture_constraints.md:34]:
```suggestion
In contrast to the list above, it is strongly recommended to use the following patterns and techniques.
```

File: markdown/02_architecture_constraints.md [hunk 4]
```diff
@@ -1,21 +1,78 @@
...
+
+| Encouraged                   | Description                                                                               | Example                        |
+| ---------------------------- | ----------------------------------------------------------------------------------------- | ------------------------------ |
+| Cache friendly containers    | Containers that barely miss L1 cache                                                      | `std::vector, std::array`      |
+| `new` allocations            | Dynamic memory allocation (malloc)                                                        | `char* chars = new char[x]`    |
+| low level abstractions       | This reduces abstraction cost                                                             | such as functions              |
```
Comment (Clazilla) [inline markdown/02_architecture_constraints.md:40]:
```suggestion
| low level abstractions       | This reduces abstraction costs                                                            | such as functions              |
```

**Comment by stritti** [comment]
Some remarks: 
- Maybe there are also some general required compiler versions/flags which has to be used in respect of the architectural decisions?
- Other architectural constraint may also be 
  - the OS (e.g. Windows 10 64Bit, no MacOS/Linux ...)
  - specific hardware (e.g. existing/specific GPU)

**Pull request merged**